File: 3871_Count_Commas_in_Range_II/sol.java
import java.util.*;

class Solution {
    /**
     * Time:    O(log N)
     * Space:   O(1)
     * Strategy: Maths
     */
    public long solve(long n) {
        long result = 0;
        for (long i = 1000; i <= n; i*=1000) {
            result += n-i +1;
        }
        return result;
    }
    public static void main(String[] args) {
        Solution sol = new Solution();
        long[] inputs = { 1002, 998, 1004590  };
        long[] expected = {3, 0, 1008182};
        System.out.println("Running Tests...");
        for (int i = 0; i < inputs.length; i++) {
            long result = sol.solve(inputs[i]);
            String status = (result == expected[i]) ? "✅ PASS" : "❌ FAIL";
            System.out.printf("%s | Test #%d: Input: %d | Expected: %d | Result: %d%n",
                status, i+1, inputs[i], expected[i], result );
        }
    }
}

File: 3871_Count_Commas_in_Range_II/sol.py
class Solution:
    """
    @param arg1: int
    @return: int
    Time:    O(log N)
    Space:   O(1)
    Strategy: int
    """
    def solve(self, n: int) -> int:
        result = 0
        i = 1000
        while i <= n:
            result += n-i +1
            i = i*1000
        return result
# ================ 🧪 Local Test Suite ======================================
def run_tests():
    sol = Solution()
    test_cases = [
        (( 1002, ), 3 ),
        (( 998, ) , 0 ),
        (( 1004590, ) , 1008182 ),
    ]
    for i, (args, expected) in enumerate(test_cases, 1):
        result = sol.solve(*args)
        status = "✅ PASS" if result == expected else "❌ FAIL"
        print(f"Test {i}: | {status} | Input={args} | Expected={expected} | Got={result}")
if __name__ == "__main__":
    run_tests()